[casaarlai] iP

casaarlai opened the pull request:
# Lai
_A rather **cynical** personality that helps you keep track of your life._
> Write it down. Written goals have a way of transforming wishes into wants; cant's into cans; dreams into plans; and plans into reality. Don't just think it - ink it! – Michael Korda [(source)](https://www.azquotes.com/quote/856710)
  
Lai is currently able to keep track of
- [x] Events (with support for start and end dates)
- [x] Deadlines (with support for due dates)
- [x] To-do items
- [ ] Emotional ramblings (hopefully Lai never has to!)

Why use Lai?
1. Completely free :star2:
2. Extremely performant 
3. New features added frequently!

Some basic commands include:
+ ```list``` to retrieve your list of tasks saved
+ ```bye``` to exit Lai
+ ```mark n``` to mark the task at the _nth_ position in the list as done
+ ```unmark n``` to mark the task at the _nth_ position in the list as not done

## Lai was proudly implemented in Java!

``` java
/**
 * Constructs a new Event with the specified description, start date, and end date.
 * The start and end dates are parsed from strings. If parsing fails using ISO-8601 format, an attempt is made using
 * "dd MMM yyyy" format.
 *
 * @param description The description of the event.
 * @param from The start date of the event in ISO-8601 format or "dd MMM yyyy" format.
 * @param to The end date of the event in ISO-8601 format or "dd MMM yyyy" format.
 * @throws DateTimeParseException If the start or end date strings cannot be parsed.
 */
public Event(String description, String from, String to) {
    super(description);
    try {
        this.from = LocalDate.parse(from);
        this.to = LocalDate.parse(to);
    } catch (DateTimeParseException e) {
        this.from = LocalDate.parse(from, DateTimeFormatter.ofPattern("dd MMM yyyy"));
        this.to = LocalDate.parse(to, DateTimeFormatter.ofPattern("dd MMM yyyy"));
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

casaarlai committed:
Changed the name of the chatbot from Duke to Lai

casaarlai committed:
Have Lai echo commands and exit on the command 'bye'

casaarlai committed:
Add ability to store and list commands entered by user

casaarlai committed:
Added Task class, and ability to mark and unmark tasks as done

casaarlai committed:
Add ToDo, Event and Deadline functionality

casaarlai committed:
Automate testing using I/O redirection

casaarlai committed:
Added error handling and LaiException

casaarlai committed:
Added error handling and LaiException

casaarlai committed:
Added support for deleting tasks

casaarlai committed:
Add ability to load/save tasks with the hard disk

casaarlai committed:
Have Lai parse input into LocalDate object

casaarlai committed:
Merge branch 'branch-Level-8'

Kappaccinoh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+        this.isDone = isDone;
+    }
+
+    public String getDone() {
```
I'm wondering if `getDone()` might be a slightly confusing naming convention, I recognise that the code here is quite simple so I can immediately tell you are retreiving the `boolean isDone` field, but perhaps renaming the function to `getisDone()` might be clearer

casaarlai replied:
@Kappaccinoh is there a reason setter functions for boolean variables follow this convention:

void setFound(boolean isFound);

but getter functions do not?

Kappaccinoh reviewed (commented):
Well done overall - I have not qualms about your naming conventions after looking through all the files. Keep up the naming conventions as you expand and build up your Lai ChatBot!

casaarlai committed:
Refactor the code to use more OOP

casaarlai committed:
Merge branch 'branch-A-MoreOOP'

casaarlai committed:
Refactor code into packages

casaarlai committed:
Merge branch 'branch-A-Packages'

casaarlai committed:
Merge remote-tracking branch 'origin/add-gradle-support'

casaarlai committed:
Add tests for Parser and Storage classes

casaarlai committed:
Merge branch 'branch-A-JUnit'

casaarlai committed:
Edit build.gradle file for packaging Lai as a JAR file

casaarlai committed:
Add JavaDoc comments for task classes

casaarlai committed:
Add JavaDoc comments for utility classes

casaarlai committed:
Tweak code to adhere to coding standard

casaarlai committed:
Add functionality for user to search for tasks with a keyword

casaarlai committed:
Merge branch 'branch-A-JavaDoc'

casaarlai committed:
Merge branch 'branch-A-CodingStandard'

casaarlai committed:
Merge branch 'branch-Level-9'

tamagochuuu commented on src/main/java/lai/task/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    /**
+     * Returns a string representation of the deadline, including its type, completion status, description, and due
+     * date.
+     *
+     * @return A string in the format "[D][completion_status] description (by: due_date)", where completion_status is
+     * "X" if the task is done, or a space character otherwise, and due_date is formatted as "dd MMM yyyy".
+     */
```
Perhaps it would be better to break some of the longer lines of comment into two lines so that they are not too long

tamagochuuu commented on src/main/java/lai/task/Event.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public String toString() {
+        DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("dd MMM yyyy");
+        return "[E]" + super.toString() + " (from: " + from.format(dateFormat) + " to: " + to.format(dateFormat) + ")";
```
Perhaps it would be best to separate this line into a two lines as it is quite long

tamagochuuu commented on src/main/java/Lai.java:
```diff
@@ -0,0 +1,117 @@
+                    if (descFrom.length > 1) {
+                        from = descFrom[1];
+                    }
+                    // Separating the end out
```
Perhaps this comment could be a little clearer

tamagochuuu reviewed (commented):
Good

E0735389 commented on src/main/java/lai/util/Storage.java:
```diff
@@ -0,0 +1,112 @@
+                } else if (line.startsWith("[E]")) {
+                    int fromIndex = description.indexOf(" (from: ");
+                    int toIndex = description.indexOf(" to: ");
+                    String fromString = description.substring(fromIndex + 8, toIndex);
```
Using `split` might be better to avoid hard coding magic constants like 8 above.

E0735389 commented on src/main/java/lai/util/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+ * as well as to iterate over all tasks.
+ */
+public class TaskList implements Iterable<Task> {
+    protected List<Task> tasks = new ArrayList<>();
```
It's not clear why the default value is needed since it's overwritten in every constructor.

E0735389 commented on src/main/java/lai/util/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+     * @return A list of {@code Task} objects whose descriptions contain the specified query.
+     */
+    public List<Task> find(String query) {
+        List<Task> result  = new ArrayList<>();
```
There should only be one space before the `=`.

ChrisHo1341 replied:
```suggestion
        List<Task> result = new ArrayList<>();
```

Might be good remove a space to follow the coding standards

E0735389 commented on src/main/java/lai/util/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+
+        Ui.printTaskAdded(newTask, this);
+
+        return this.tasks;
```
Be careful, the caller may accidentally modify the internal variable `tasks` if you do this.

E0735389 commented on src/main/java/lai/util/Ui.java:
```diff
@@ -0,0 +1,142 @@
+    /**
+     * Prints a decorative line to separate sections of output for better readability.
+     */
+    public static void printLine() {
```
Consider "printHorizontalLine" or "printDecorativeLine". Since println in fact stands for printline.

ChrisHo1341 commented on src/main/java/lai/util/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+     * @param storage The storage utility to update the tasks file.
+     * @return The updated list of tasks.
+     */
+    public List<Task> add(Task newTask, Storage storage) {
```
Consider renaming to `addTask() `to be even clearer

ChrisHo1341 commented on src/main/java/lai/util/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+     * @param index The index of the task in the list.
+     * @return The task at the specified index.
+     */
+    public Task get(int index) {
```
Consider renaming the function to `getTask()` to make it more clear

ChrisHo1341 reviewed (commented):
I found the code to be easy to read. Some function names are a bit ambiguous, consider changing them to make them clearer

casaarlai committed:
Added MainWindow and DialogBox for GUI

casaarlai committed:
Merge branch 'branch-Level-10'

casaarlai committed:
Removed unneeded method calls

casaarlai committed:
Add assert statements

The program has no assert statements at all.

Assert statements help to define important assumptions about the
program state so that the runtime can verify them. The program can also
then stop the execution sooner if there is a possible bug which is
safer.

Add assert statements at points where crucial
assumptions are made in the program

Such points in the program where crucial assumptions are made are more
likely to cause further issues during execution and should be
prioritized.

casaarlai committed:
Refactor to improve code quality

The main method getResponse is too long and has multiple levels of
abstraction

Extract out methods based on what command they are invoked by

This will allow the getResponse method to have a single level of
abstraction (SLAP) that makes the method much easier to read and
understand

Refer to this for an explanation on SLAP:
https://nus-cs2103-ay2324s2.github.io/website/book/codeQuality/maximizeReadability/intermediate/slapHard/

casaarlai committed:
Merge pull request #2 from casaarlai/branch-A-CodeQuality

Refactor to improve code quality

casaarlai committed:
Merge branch 'master' into branch-A-Assertions

casaarlai committed:
Merge pull request #3 from casaarlai/branch-A-Assertions

Add assert statements

casaarlai committed:
Add function to detect if any duplicates exist before a task is being added

casaarlai committed:
Update JavaDoc

casaarlai committed:
Merge pull request #4 from casaarlai/branch-C-DetectDuplicates

Detect duplicate tasks

casaarlai committed:
Edit user guide

casaarlai committed:
Merge pull request #5 from casaarlai/A-UserGuide

Edit user guide

casaarlai committed:
Fix link for user guide image

casaarlai committed:
Edit FXML for DialogBox to prevent text from cutting off
